File: csharp/1156.swap-for-longest-repeated-character-substring.cs
/*
 * @lc app=leetcode id=1156 lang=csharp
 *
 * [1156] Swap For Longest Repeated Character Substring
 */

using System;

// @lc code=start
public class Solution {
    public int MaxRepOpt1(string text) {
        int n = text.Length, ret;
        int[] count = new int[n], total = new int[26];
        count[0] = 1;
        total[(int)text[0] - (int)'a'] = 1;
        for (int i = 1; i < n; ++i) {
            if (text[i] == text[i - 1]) {
                count[i] = count[i - 1] + 1;
            } else {
                count[i] = 1;
            }
            ++total[(int)text[i] - (int)'a'];
        }
        ret = count[n - 1] + (total[(int)text[n - 1] - (int)'a'] > count[n - 1] ? 1 : 0);
        for (int i = n - 2; i >= 0; --i) {
            if (text[i] == text[i + 1]) {
                count[i] = count[i + 1];
            } else {
                if (count[i] == total[(int)text[i] - (int)'a']) {
                    ret = Math.Max(ret, count[i]);
                } else {
                    ret = Math.Max(ret, 1 + count[i]);
                }
            }
        }
        for (int i = 1; i < n - 1; ++i) {
            if (text[i] != text[i - 1] && text[i] != text[i + 1] && text[i - 1] == text[i + 1]) {
                int b = count[i - 1] + count[i + 1];
                if (total[(int)text[i - 1] - (int)'a'] > b) {
                    ++b;
                }
                ret = Math.Max(b, ret);
            }
        }
        return ret;
    }
}
// @lc code=end

File: java/1156.swap-for-longest-repeated-character-substring.java
/*
 * @lc app=leetcode id=1156 lang=java
 *
 * [1156] Swap For Longest Repeated Character Substring
 */

// @lc code=start
class Solution {
    public int maxRepOpt1(String text) {
        int i = 0, n = text.length(), pp = 0, p = 0, ret = 1;
        char ppc = '\0', pc = '\0';
        int[] maxSegmentWithGap = new int[26];
        int[] maxSegment = new int[26];
        int[] firstIndexPlusOne = new int[26];
        while (i < n) {
            char c = text.charAt(i);
            int k = c - 'a', j;
            ret = Math.max(Math.max(ret, 1 + maxSegment[k]), 1 + maxSegmentWithGap[k]);
            for (j = i + 1; j < n && text.charAt(j) == c; ++j);
            if (ppc == c && p == 1) {
                ret = Math.max(ret, pp + j - i + (firstIndexPlusOne[k] != 0 && firstIndexPlusOne[k] < i - 1 - pp ? 1 : 0));
                maxSegmentWithGap[k] = Math.max(maxSegmentWithGap[k], pp + j - i);
            } else {
                ret = Math.max(ret, j - i + (firstIndexPlusOne[k] != 0 ? 1 : 0));
            }
            pp = p;
            ppc = pc;
            p = j - i;
            pc = c;
            if (firstIndexPlusOne[k] == 0) {
                firstIndexPlusOne[k] = i + 1;
            }
            maxSegment[k] = Math.max(maxSegment[k], j - i);
            i = j;
        }
        return ret;
    }
}
// @lc code=end

